Login MFA docs (#14317)
hashicorp/vault: website/content/docs/auth/login-mfa.mdx

---
layout: docs
page_title: Multi-Factor Authentication (MFA) for Login - Auth Methods
description: |-
  Multi-factor authentication (MFA) is supported for several authentication
  methods.
---
# Login MFA

Vault has support for Multi-factor Authentication (MFA) for authenticating to
an auth method, using different authentication types. We use the term `Login MFA` to distinguish
this feature and the [Vault Enterprise MFA](/docs/enterprise/mfa).
Login MFA is built on top of the Identity system of Vault.

## MFA Types

Login MFA in Vault can be of the following types.

- `Time-based One-time Password (TOTP)` - If configured and enabled on a login path,
  this would require a TOTP passcode along with a Vault token, to be presented
  while invoking the API login request. The passcode will be validated against the
  TOTP key present in the identity of the caller in Vault.

- `Okta` - If Okta push is configured and enabled on a login path, then the enrolled
  device of the user will get a push notification to approve or deny the access
  to the API. The Okta username will be derived from the caller identity's
  alias.

> - `Duo` - If Duo push is configured and enabled on a login path, then the enrolled

Reviewer: should we note that Duo validation can also be done via passcode ?

  device of the user will get a push notification to approve or deny the access
  to the API. The Duo username will be derived from the caller identity's
  alias.

- `PingID` - If PingID push is configured and enabled on a login path, then the
  enrolled device of the user will get a push notification to approve or deny
  the access to the API. The PingID username will be derived from the caller
  identity's alias.

## Login MFA Procedure

Login MFA can be configured to further secure authenticating to an auth method. To enable login
MFA, an MFA method needs to be configured. Please see [Login MFA API](/api/secret/identity/mfa) for details
on how to configure an MFA method. Once an MFA method is configured, using the returned unique MFA method ID,
an operator can configure an MFA enforcement. Please see [Login MFA Enforcement API](/api/secret/identity/mfa/login-enforcement)
for details on how to configure an MFA enforcement config. MFA could be enforced for an entity, a group of
entities, a specific auth method accessor, or an auth method type. A login request that matches

> either of the MFA enforcement restrictions is subject to further MFA validation,

Reviewer: ```suggestion
any of the MFA enforcement restrictions is subject to further MFA validation,
```

like a one-time passcode, before being authenticated.

There are two ways to validate a login request that is subject to MFA validation.

 ### Single-Phase Login
 In the Single-phase login, the required MFA information is embedded in a login request using
 the `X-Vault-MFA` header. In this case, the MFA validation is done
 as a part of the login request.

 MFA credentials are retrieved from the `X-Vault-MFA` HTTP header. The format of
 the header is `mfa_method_id[:key[=value]]`. The item in the `[]` is optional. If there are more than
 one MFA method that needs to be validated, a user can pass in multiple `X-Vault-MFA` HTTP headers.

 #### Sample Request

 ```shell-session
 $ curl \
     --header "X-Vault-Token: ..." \

>      --header "X-Vault-MFA:d16fd3c2-50de-0b9b-eed3-0301dadeca10:695452" \

Reviewer: should there be a space here between header key and value, just for visual consistency with X-Vault-Token above i.e "X-Vault-MFA: d16fd3c2-50de-0b9b-eed3-0301dadeca10:695452" ?

     http://127.0.0.1:8200/v1/auth/userpass/login/alice
 ```

 ### Two-Phase Login
 The more conventional, and prevalent MFA method is a two request mechanism, also referred to as Two-phase Login MFA.
 In Two-phase login the `X-Vault-MFA` header is not provided in the request. In this case, after sending a regular login request,
 the user receives an auth response in which MFA requirements are included. MFA requirements contain an MFA request ID
 which identifies the login request that needs to get validated. In addition, MFA requirements contain MFA constraints
 that identify which MFA types should be used to validate the request, the corresponding method IDs, and
 a boolean value showing whether the MFA method uses passcodes or not. MFA constraints form a nested map in MFA Requirement,

>  and represent all MFA enforcements that match a login request.

Reviewer: Would something like this make sense, just to call out that we aren't updating all login docs?
`While the example below is for the userpass login, it should be noted that this can affect the login response on any auth mount protected by MFA validation`

#### Sample Two-Phase Login Response

```json
{
  "request_id": "1044c151-13ea-1cf5-f6ed-000c42efd477",
  "lease_id": "",
  "lease_duration": 0,
  "renewable": false,
  "data": null,
  "warnings": [
    "A login request was issued that is subject to MFA validation. Please make sure to validate the login by sending another request to mfa/validate endpoint."
  ],
  "auth": {
    "client_token": "",
    "accessor": "",
    "policies": null,
    "token_policies": null,
    "identity_policies": null,
    "metadata": null,
    "orphan": false,
    "entity_id": "",
    "lease_duration": 0,
    "renewable": false,
    "mfa_requirement": {
      "mfa_request_id": "d0c9eec7-6921-8cc0-be62-202b289ef163",
      "mfa_constraints": {
        "enforcementConfigUserpass": {
          "any": [
            {
              "type": "totp",
              "id": "820997b3-110e-c251-7e8b-ff4aa428a6e1",
              "uses_passcode": true
            }
          ]
        }
      }
    }
  }
}
```

 Note that the `uses_passcode` boolean value is always set to true for TOTP, and always set to false for Okta and PingID.

>  For DUO method, the value can be configured as part of the method configuration.

Reviewer 2: In some places, I see `DUO` and others `Duo`. Duo's web site uses `Duo`, so maybe we could use that everywhere?

 Please see [DUO API](/api/secret/identity/mfa/duo) for details
 on how to configure the boolean value for DUO.

>  To validate the MFA restricted login request, the user sends a second request to the [validate](/sys/mfa/validate)

Reviewer 2: This link doesn't work.

 endpoint including the MFA request ID, and MFA payload. MFA payload contains a map of methodIDs and their associated credentials.
 If the configured MFA methods (like PingID, Okta, Duo) don't need a passcode, the associated

>  credentials will be a list of an empty string.

Reviewer: ```suggestion
 credentials will be a list with one empty string.
```
at a glance, the length of the list seemed a bit unclear, hence dropping the suggestion

  TOTP key present in the caller's identify in Vault.

- `Okta` - If Okta push is configured and enabled on a login path, then the enrolled
  device of the user will receive a push notification to either approve or deny  access
  to the API. The Okta username will be derived from the caller identity's
  alias.

- `Duo` - If Duo push is configured and enabled on a login path, then the enrolled
  device of the user will receive a push notification to either approve or deny access
  to the API. The Duo username will be derived from the caller identity's
  alias. Note that Duo could also be configured to use passcodes for authentication.

- `PingID` - If PingID push is configured and enabled on a login path, then the
  enrolled device of the user will receive a push notification to either approve or deny
access to the API. The PingID username will be derived from the caller
  identity's alias.

## Login MFA Procedure

Login MFA can be configured to further secure authenticating to an auth method. To enable login
MFA, an MFA method needs to be configured. Please see [Login MFA API](/api/secret/identity/mfa) for details

> on how to configure an MFA method. Once an MFA method is configured, using the returned unique MFA method ID,

Reviewer 3: ```suggestion
on how to configure an MFA method. Once an MFA method is configured, an operator can configure an MFA enforcement using the returned unique MFA method ID.
```

an operator can configure an MFA enforcement. Please see [Login MFA Enforcement API](/api/secret/identity/mfa/login-enforcement)
for details on how to configure an MFA enforcement config. MFA could be enforced for an entity, a group of
entities, a specific auth method accessor, or an auth method type. A login request that matches

> any of the MFA enforcement restrictions is subject to further MFA validation,

Reviewer 3: ```suggestion
any MFA enforcement restrictions is subject to further MFA validation,
```

> like a one-time passcode, before being authenticated.

Reviewer 3: ```suggestion
such as a one-time passcode, before being authenticated.
```

> Although an MFA method configuration like (Okta, Duo, and PingID) needs a mount accessor and an MFA enforcement configuration

Reviewer 3: ```suggestion
Although an MFA method configuration , such as Okta, Duo, and PingID, needs a mount accessor and an MFA enforcement configuration
```

> could use mount accessors. However, there is no particular correlation between the two accessors. In other words, the

Reviewer: Should we clarify the "no particular correlation" part here a bit? Since there does need to be some relation to map the user trying to login to the accessor on the method, iiuc. Maybe something like
```The accessors do not necessarily need to be the same for the 2 cases. However, the user trying to login on an MFA enforced mount would need to have an alias on the accessor in the MFA method itself, in the case that entity/alias templating is required to render out a username``` ?

> However, a user trying to login on an MFA enforced mount would need to have an alias on the accessor in the MFA method itself.

Reviewer 3: ```suggestion
However, a user trying to log in on an MFA enforced mount needs to have an alias on the accessor in the MFA method itself.
```

> However, a user trying to login on an MFA enforced mount would need to have an alias on the accessor in the MFA method itself.

Reviewer 3: ```suggestion
However, a user trying to login on an MFA enforced mount needs to have an alias on the accessor in the MFA method itself.
```

> In this case, the entity/alias templating is required to render out the username associated third party account.

Reviewer 3: ```suggestion
In this case, the entity/alias templating must render out the username associated third party account.
```

Reviewer 3: Do you mean: In this case, the entity/alias templating must render out the username associated **with the** third party account.

There are two ways to validate a login request that is subject to MFA validation.

 ### Single-Phase Login

>  In the Single-phase login, the required MFA information is embedded in a login request using

Reviewer 3: ```suggestion
 In the Single-phase login, the required MFA information is embeds in a login request using
```

 as a part of the login request.

 MFA credentials are retrieved from the `X-Vault-MFA` HTTP header. The format of

>  the header is `mfa_method_id[:passcode]`. The item in the `[]` is optional. If there are more than
>  one MFA method that needs to be validated, a user can pass in multiple `X-Vault-MFA` HTTP headers.

Reviewer: ```suggestion
 the header is `mfa_method_id[:passcode]`. The item in the `[]` is optional. If there are multiple MFA methods that need to be validated, a user can pass in multiple `X-Vault-MFA` HTTP headers.
```

 #### Sample Request

 ```shell-session
 $ curl \
     --header "X-Vault-Token: ..." \
     --header "X-Vault-MFA: d16fd3c2-50de-0b9b-eed3-0301dadeca10:695452" \
     http://127.0.0.1:8200/v1/auth/userpass/login/alice
 ```

 If an MFA method does not require passcode, the login request MFA header should only contain the method ID.

 ```shell-session
  $ curl \
      --header "X-Vault-Token: ..." \
      --header "X-Vault-MFA: d16fd3c2-50de-0b9b-eed3-0301dadeca10" \
      http://127.0.0.1:8200/v1/auth/userpass/login/alice
  ```

 ### Two-Phase Login

>  The more conventional, and prevalent MFA method is a two request mechanism, also referred to as Two-phase Login MFA.

Reviewer 3: ```suggestion
 The more conventional and prevalent MFA method is a two-request mechanism, also referred to as Two-phase Login MFA.
```

>  In Two-phase login the `X-Vault-MFA` header is not provided in the request. In this case, after sending a regular login request,

Reviewer 3: ```suggestion
 In Two-phase login, the `X-Vault-MFA` header is not provided in the request. In this case, after sending a regular login request,
```

 the user receives an auth response in which MFA requirements are included. MFA requirements contain an MFA request ID

>  which identifies the login request that needs to get validated. In addition, MFA requirements contain MFA constraints

Reviewer 3: ```suggestion
 which identifies the login request that needs validation. In addition, MFA requirements contain MFA constraints
```

>  that identify which MFA types should be used to validate the request, the corresponding method IDs, and

Reviewer 3: ```suggestion
 that determine which MFA types should be used to validate the request, the corresponding method IDs, and
```

>  a boolean value showing whether the MFA method uses passcodes or not. MFA constraints form a nested map in MFA Requirement,

Reviewer 3: ```suggestion
 a boolean value showing whether the MFA method uses passcodes or not. MFA constraints form a nested map in MFA Requirement
```

>  and represent all MFA enforcements that match a login request. While the example below is for the userpass login, it should

Reviewer 3: ```suggestion
 and represent all MFA enforcements that match a login request. While the example below is for the userpass login, 
```

>  be noted that this can affect the login response on any auth mount protected by MFA validation.

Reviewer 3: ```suggestion
 note that this can affect the login response on any auth mount protected by MFA validation.
```

  ],
  "auth": {
    "client_token": "",
    "accessor": "",
    "policies": null,
    "token_policies": null,
    "identity_policies": null,
    "metadata": null,
    "orphan": false,
    "entity_id": "",
    "lease_duration": 0,
    "renewable": false,
    "mfa_requirement": {
      "mfa_request_id": "d0c9eec7-6921-8cc0-be62-202b289ef163",
      "mfa_constraints": {
        "enforcementConfigUserpass": {
          "any": [
            {
              "type": "totp",
              "id": "820997b3-110e-c251-7e8b-ff4aa428a6e1",
              "uses_passcode": true
            }
          ]
        }
      }
    }
  }
}
```

 Note that the `uses_passcode` boolean value is always set to true for TOTP, and always set to false for Okta and PingID.
 For Duo method, the value can be configured as part of the method configuration.
 Please see [Duo API](/api/secret/identity/mfa/duo) for details
 on how to configure the boolean value for Duo.

 To validate the MFA restricted login request, the user sends a second request to the [validate](/api/system/mfa/validate)

>  endpoint including the MFA request ID, and MFA payload. MFA payload contains a map of methodIDs and their associated credentials.

Reviewer 3: ```suggestion
 endpoint including the MFA request ID and MFA payload. MFA payload contains a map of methodIDs and their associated credentials.
```

>  If the configured MFA methods (like PingID, Okta, Duo) don't need a passcode, the associated

Reviewer 3: ```suggestion
 If the configured MFA methods, such as PingID, Okta, and Duo, do not require a passcode, the associated
```

 credentials will be a list with one empty string.

#### Sample Payload

```json
{
  "mfa_request_id": "5879c74a-1418-1948-7be9-97b209d693a7",
  "mfa_payload": {
      "d16fd3c2-50de-0b9b-eed3-0301dadeca10": ["910201"]
  }
}
```

#### Sample Request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request POST \
    --data @payload.json \
    http://127.0.0.1:8200/v1/sys/mfa/validate
```

#### Sample CLI Request
A user is also able to use the CLI write command to validate the login request.

```shell-session
$ vault write sys/mfa/validate -format=json @payload.json
```

#### Interactive CLI for Login MFA
Vault supports an interactive way of authenticating to an auth method using CLI only if the
login request is subject to a single MFA method validation. In this situation, if the MFA method
is configured to use passcodes, after sending a regular login request, the user is prompted to
insert the passcode. Upon successful MFA validation, a client token is returned.

> If the configured MFA methods (like PingID, Okta, Duo) don't need a passcode and have out of band

Reviewer 3: ```suggestion
If the configured MFA methods, such as PingID, Okta, and Duo, do not require a passcode and have out of band
```

---
layout: docs
page_title: Multi-Factor Authentication (MFA) for Login - Auth Methods
description: |-
  Multi-factor authentication (MFA) is supported for several authentication
  methods.
---
# Login MFA

> Vault has support for Multi-factor Authentication (MFA) for authenticating to

Reviewer 3: ```suggestion
Vault supports Multi-factor Authentication (MFA) for authenticating to
```

> an auth method, using different authentication types. We use the term `Login MFA` to distinguish

Reviewer 3: ```suggestion
an auth method using different authentication types. We use the term `Login MFA` to distinguish
```

Reviewer 3: Recommendation: when referencing terms, use italics instead of wrapping the term in code.

this feature and the [Vault Enterprise MFA](/docs/enterprise/mfa).
Login MFA is built on top of the Identity system of Vault.

## MFA Types

> Login MFA in Vault can be of the following types.

Reviewer 3: ```suggestion
MFA in Vault includes the following login types:
```

- `Time-based One-time Password (TOTP)` - If configured and enabled on a login path,

>   this would require a TOTP passcode along with a Vault token, to be presented

Reviewer 3: ```suggestion
  this would require a TOTP passcode along with a Vault token to be presented
```

  while invoking the API login request. The passcode will be validated against the

>   TOTP key present in the identity of the caller in Vault.

Reviewer 3: ```suggestion
  TOTP key present in the caller's identify in Vault.
```

- `Okta` - If Okta push is configured and enabled on a login path, then the enrolled

>   device of the user will get a push notification to approve or deny the access

Reviewer 3: ```suggestion
  device of the user will receive a push notification to either approve or deny the access
```

  to the API. The Okta username will be derived from the caller identity's
  alias.

- `Duo` - If Duo push is configured and enabled on a login path, then the enrolled

>   device of the user will get a push notification to approve or deny the access

Reviewer 3: ```suggestion
  device of the user will receive a push notification to either approve or deny the access
```

  to the API. The Duo username will be derived from the caller identity's
  alias. Note that Duo could also be configured to use passcodes for authentication.

- `PingID` - If PingID push is configured and enabled on a login path, then the

>   enrolled device of the user will get a push notification to approve or deny

Reviewer 3: ```suggestion
  enrolled device of the user will receive a push notification to either approve or deny
```

>   the access to the API. The PingID username will be derived from the caller

Reviewer 3: ```suggestion
access to the API. The PingID username will be derived from the caller
```

  identity's alias.

## Login MFA Procedure

> Login MFA can be configured to further secure authenticating to an auth method. To enable login

Reviewer 3: **Recommendation**: Try and avoid spitting infinitives: '...to secure' has been split by the infinitive 'further'. 
**Suggestion**: Login MFA can be configured to secure further authenticating to an auth method.
